### PR TITLE
[XLA] Add optional call stack to OpMetaData in xla_data proto

### PR DESCRIPTION
**Pull request by mrnikwaws**
When profiling a compiled model using XLA it can be useful to examine the call stack (file paths and line numbers) which contributed to a lowered operator.  Where a non-performant lowering is due to the structure or parameterization in the model code, a customer is then empowered to find the various relevant points in the call stack that contributed.

This change extends the existing file number and line number entries in the op meta data (xla_data.proto), to (optionally) include the full stack of locations.  "call_stack" is added as a parent to the repeated path, line number pairs for clarity.

### CODE REVIEW

**Comment by ddunl** [comment]
This seems good to me, but I'm not really the proper person to review this. I'll wait for @tpopp to review as well. Thank you for your PR!!

**Comment by hawkinsp** [comment]
I like this idea in general: JAX has wanted to emit full stack traces in MHLO and HLO for some time, but there has been no representation of them so we haven't been able to.

There are a number of things that we'll have to think about here:
a) what's the printed representation of stack frames? HLO can also be serialized as text.
a) stack traces can be very big: you might easily have dozens of frames on every operator, and if not done carefully this will blow up the size of the HLO representation both in memory and when serialized.

However stack traces are also highly redundant: they often have common prefixes, and they often share strings. I think we would need a more compressed representation (e.g., a shared DAG of stack trace nodes for the whole module, or a string table, or both.)

File: xla/xla_data.proto [hunk 1]
```diff
@@ -383,6 +383,20 @@ message OpMetadata {
...
+    // Stack frame containing file and line number relating to the operator
+    message CallStackFrame {
+      string source_file = 1;
+      int32 source_line = 2;
+    }
+    repeated CallStackFrame frame = 1;
+  }
+
+  // Call stack info for operator lowering
+  OpCallStack call_stack = 13;
```
Comment (cheshire) [inline xla/xla_data.proto:399]: Most stack traces share subtraces, so there needs to be a level of indirection there.

**Comment by sherhut** [comment]
A source of inspiration could be how this is handled in MLIR (and mhlo): Attach a table of locations that can reference themselves to the module and only reference entries from the table from hlo operations. 
We probably only need few message types. A leaf node like what is proposed here and a call-site node might be all that is needed.

**Comment by mrnikwaws** [comment]
> This makes sense; do you plan to wire up the framework integration?

Yes - once we arrive at an agreed HLO.

**Comment by mrnikwaws** [comment]
> I like this idea in general: JAX has wanted to emit full stack traces in MHLO and HLO for some time, but there has been no representation of them so we haven't been able to.
> 
> There are a number of things that we'll have to think about here: a) what's the printed representation of stack frames? HLO can also be serialized as text. b) stack traces can be very big: you might easily have dozens of frames on every operator, and if not done carefully this will blow up the size of the HLO representation both in memory and when serialized.
> 
> However stack traces are also highly redundant: they often have common prefixes, and they often share strings. I think we would need a more compressed representation (e.g., a shared DAG of stack trace nodes for the whole module, or a string table, or both.)

This is good feedback.  We are currently looking at a simple tree structured string table (e.g. a message with a parent_id and path segment string in a table, where only root nodes have no parent).  This would allow a single index in the "source_file" field pointing at the leaf node. I don't think a more elaborate DAG would provide much value in this case.

I think this should allow the data to be populated quickly and in a single pass, so we don't want to add too much overhead to the lowering process.

**Comment by mrnikwaws** [comment]
Code is now updated to show the proposed solution.  In essence we use a tree of path segments, and the `source_file_as_path_segment_id` points to the leaf node of a path so storage remains efficient in each call stack frame too

**Comment by mrnikwaws** [comment]
> A source of inspiration could be how this is handled in MLIR (and mhlo): Attach a table of locations that can reference themselves to the module and only reference entries from the table from hlo operations. We probably only need few message types. A leaf node like what is proposed here and a call-site node might be all that is needed.

I didn't quite grok this, and I am not familiar with MHLO - however specific feedback on how the proposed change can be improved is welcome

File: xla/xla_data.proto (hunk 1 above)
Comment (mrnikwaws) [reply xla/xla_data.proto]: I think this should be good now

**Comment by alanwaketan** [comment]
Second this idea.

**Comment by sherhut** [comment]
> I didn't quite grok this, and I am not familiar with MHLO - however specific feedback on how the proposed change can be improved is welcome

You already proposed a similar design. As we might need/want to load this data or produce it from an MLIR based abstraction (like mhlo and stablehlo), my concern was to chose a format that can easily be expressed in those frameworks. I think what you have proposed here could be encoded there.

MLIR has built-in support to unique attributes in the IR and location information uses the same approach. Instead of centering around string and a path (which is unstructured), the MLIR encoding has specific location types. The leaf one in this case would be a file location (see https://github.com/llvm/llvm-project/blob/main/mlir/include/mlir/IR/BuiltinLocationAttributes.td#L65) and a callsite location (see https://github.com/llvm/llvm-project/blob/main/mlir/include/mlir/IR/BuiltinLocationAttributes.td#L27) allows to generated a stack out of it.

When printing (or transforming into proto), these also need to turn into a table and are referenced by ids. 

To enable easy interop, we can give the proto based implementation the same structure (and model file positions and call sites) or, as an alternative, implement the unstructured locations in mlir. I would vote for the former, given that we already encode file position information.

Also @burmako FYI and for your input on StableHLO's perspective.

**Comment by mrnikwaws** [comment]
> > I didn't quite grok this, and I am not familiar with MHLO - however specific feedback on how the proposed change can be improved is welcome
> 
> You already proposed a similar design. As we might need/want to load this data or produce it from an MLIR based abstraction (like mhlo and stablehlo), my concern was to chose a format that can easily be expressed in those frameworks. I think what you have proposed here could be encoded there.
> 
> MLIR has built-in support to unique attributes in the IR and location information uses the same approach. Instead of centering around string and a path (which is unstructured), the MLIR encoding has specific location types. The leaf one in this case would be a file location (see https://github.com/llvm/llvm-project/blob/main/mlir/include/mlir/IR/BuiltinLocationAttributes.td#L65) and a callsite location (see https://github.com/llvm/llvm-project/blob/main/mlir/include/mlir/IR/BuiltinLocationAttributes.td#L27) allows to generated a stack out of it.
> 
> When printing (or transforming into proto), these also need to turn into a table and are referenced by ids.
> 
> To enable easy interop, we can give the proto based implementation the same structure (and model file positions and call sites) or, as an alternative, implement the unstructured locations in mlir. I would vote for the former, given that we already encode file position information.
> 
> Also @burmako FYI and for your input on StableHLO's perspective.

Note to self - I need to study HLO lowering in detail :).

Looking at the source supplied I am assuming there is not existing top level proto message to reference.  It seems the asks are:

* Add a "column" value (in addition to path and line)
* In addition to a "parent"/"caller" (i.e. pointer toward the base of the stack) we also have a "child"/"callee" - akin to a doubly linked list

My reading is - if those are present then lowering is easier by virtue of using existing location types.  Please confirm this is correct!  

Please also LMK if there is a preferred naming for fields (I am mostly agnostic). Alternatively, if there is an existing proto message I can co-opt - I am happy to do so - just point me to it.

I'm open to all of these, but I note it pushes us in the opposite direction wrt to a concise encoding.  

The column info is certainly present in the Python C-API so we can extract the info - I can see this aids debugging in cases where a single line of code is overloaded (it is also a convention for call stacks).  

Similarly, we can also add "callee" though I wonder at the value of this addition as the information is implicit in a call stack (i.e. a single pointer / handle is sufficient in this case).  It might be we gain reuse at a lower layer at the cost of an inflated serialization format - unless we selectively leave it unpopulated.

**Comment by mrnikwaws** [comment]
@sherhut @hawkinsp if you can confirm this approach is closer to what we want I'll get updated changes up

**Comment by tdanyluk** [comment]
As part of the OpenXLA PR rotation:
@hawkinsp advised me to refer to @cheshire instead of him, as @cheshire created an internal bug (b/283456098) for this, implying that he's considering this a part of his teams scope.

**Comment by sherhut** [comment]
This slipped by review queue. 

I am not convinced that the path segments are important, as long as we share the strings for the path, i.e., a 'FileLocation' references a string table and does not encode a string directly. This would mean that we encode each file at most once and I would not expect that many files to be in play that this becomes an issue.

Instead, we should store `StackTraceSegments` so that we can share common prefixes of stack traces themselves.

In the mlir encoding this is done via the `CallSite` location, which can be recursive. As these are all unique, a common prefix of a stack trace will only have a single representation in memory. This would be taken care of by the `StackTraceSegments` encoding.

So, having an encoding that stores all the information present in `FileLocation`, only stores strings once and has a way to represent common stack trace prefixes should be easily convertible between mlir and proto and be quite efficient in storage space.

**Comment by cheshire** [comment]
Would it be useful to add some tests, both to demonstrate the representation, show the lack of duplication, as well as mhlo/stablehlo interconversion?

**Comment by radhakrishnaba** [comment]
Hi @mrnikwaws, Is there any further update on this PR? Thank you.

**Comment by radhakrishnaba** [comment]
Hi @mrnikwaws, closing this for now. Feel free to re-open if you are working on this later.  Thanks.

**Pull request closed without merging**